MINOR: Adjust the timing for creating connect config

majialoong opened the pull request:
[In this PR](https://github.com/apache/kafka/pull/20334), we added some
validation checks for the connect config, such as ensuring that
`plugin.path` cannot be empty.

 However, currently, Connect first loads the plugin and then creates the
configuration. Even if `plugin.path` is empty, it still attempts to load
the plugin first, and then throws an exception when creating the
configuration.

The approach should be to first create a configuration to validate that
the config meet the requirements, and then load the plugin only if the
validation passes. This allows for early detection of problems and
avoids unnecessary plugin loading processes.

Reviewers: Ken Huang <s7133700@gmail.com>, Chia-Ping Tsai
 <chia7712@gmail.com>

majialoong committed:
MINOR: Adjust the timing for creating connect config

Yunyung commented:
Thanks for the PR. It looks like this also happened for MirrorMaker.

majialoong committed:
Additional adjust to MirrorMaker

majialoong commented:
@Yunyung Thank you for pointing this out, MirrorMaker has also been adjusted.

majialoong committed:
Adjust the warning log and add non-empty check

majialoong committed:
Restore code style

majialoong commented:
Hello, @Yunyung @chia7712 @m1a2st , when have time, please review this PR , thanks !

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginUtils.java:
```diff
@@ -209,7 +209,7 @@ public static Set<Path> pluginLocations(String pluginPath, boolean failFast) {
             try {
                 Path pluginPathElement = Paths.get(path).toAbsolutePath();
-                if (pluginPath.isEmpty()) {
+                if (path.isEmpty()) {
```
Is this a bug fix?

majialoong replied:
First, [in this PR](https://github.com/apache/kafka/pull/20334), `plugin.path` is defined as a non-empty parameter, so this conditional statement should never be executed.

Second, I think this conditional statement was ambiguous. The warning log indicates that the plugin path element is empty, but the actual check is for the entire plugin path, not just the element.

Therefore, we can either remove this conditional statement（since `plugin.path` cannot be empty）, or modify it to better match the warning log message about an empty plugin path element. I prefer the latter approach.

m1a2st replied:
I think we can remove this check, the `path.isEmpty()` statement will also never be executed. The `List.validator` already validates empty values in the list.

https://github.com/apache/kafka/blob/423330ebe7d76697f6cb108530fcd4cb8be1db48/clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java#L1063

m1a2st replied:
I think we should also update the `ConnectPluginPath` class to validate that `plugin.path` is not empty. However, this change might require a separate KIP or an update to KIP-1161 to address it.

majialoong replied:
> I think we should also update the `ConnectPluginPath` class to validate that `plugin.path` is not empty. However, this change might require a separate KIP or an update to KIP-1161 to address it.

Thanks for your suggestions. I will submit a separate patch to fix the ConnectPluginPath issue; this patch removes the corresponding changes.

chia7712 commented on tools/src/main/java/org/apache/kafka/tools/ConnectPluginPath.java:
```diff
@@ -171,7 +171,7 @@ private static Set<Path> parseLocations(ArgumentParser parser, Namespace namespa
             }
             String pluginPath = properties.getProperty(WorkerConfig.PLUGIN_PATH_CONFIG);
-            if (pluginPath != null) {
+            if (pluginPath != null && !pluginPath.isEmpty()) {
```
ditto

majialoong replied:
[In this PR](https://github.com/apache/kafka/pull/20334), `plugin.path` is defined to not be empty. Therefore, I think some adjustments are needed here to align with the definition of "non-empty."

majialoong replied:
@chia7712 If you think these two changes shouldn't be included in this PR I can make them separately in another PR.

chia7712 replied:
yes, it would be nice to have a separate PR to revisit those checks

majialoong committed:
Remove the modify related to ConnectPluginPath

gharris1727 commented:
@chia7712 @majialoong I don't think this change is correct.

The instantiation of the config must happen after the call to `plugins.compareAndSwapWithDelegatingLoader();`, otherwise plugin classes which are referenced within the body of the worker config but only present within the plugin.path won't be loadable.

e.g. a custom override policy is installed on the plugin.path and not the classpath, and is no longer able to be found during worker startup, or is initialized with the wrong thread context class loader.

chia7712 commented:
> a custom override policy is installed on the plugin.path and not the classpath, and is no longer able to be found during worker startup, or is initialized with the wrong thread context class loader.

oh, you are right. @majialoong would you mind opening a PR to revert this change? Also, could you please add either a test or a comment to prevent this error from happening again?

majialoong commented:
@gharris1727 I apologize for causing this error. I only focused on validating the `plugin.path` parameter and ignored the impact of other parameters. Thank you for pointing this out; I now understand the importance of this order.

I have reverted these changes in another PR https://github.com/apache/kafka/pull/20891 and implemented unit tests to prevent the issue from recurring. 

Could you please review it when you have time. Thank you again for pointing out the problem, and I apologize for the changes.
